### PR TITLE
♻️ Raise Domain Events in the Domain Layer

### PR DESCRIPTION
**Pull request by danielmackay**

### CODE REVIEW

File: src/Infrastructure/Persistence/Configuration/TodoItemConfiguration.cs [hunk 1]
```diff
@@ -12,10 +12,11 @@ public class TodoItemConfiguration : IEntityTypeConfiguration<TodoItem>
     public void Configure(EntityTypeBuilder<TodoItem> builder)
     {
         builder.HasKey(t => t.Id);
+
         builder.Property(t => t.Id)
             .HasConversion(x => x.Value,
                 x => new TodoItemId(x))
-            .ValueGeneratedOnAdd();
+            .ValueGeneratedNever(); // TODO DM: Discuss this with SAW team
```
Comment (danielmackay) [inline src/Infrastructure/Persistence/Configuration/TodoItemConfiguration.cs:19]: Would be keen to get feedback on this.  This is required if we want to manually set GUIDs used for IDs  manually in code.

File: tests/Domain.UnitTests/Entities/TodoItemTests.cs [hunk 2]
```diff
@@ -1,14 +1,75 @@
 using SSW.CleanArchitecture.Domain.Entities;
+using SSW.CleanArchitecture.Domain.Events;
 
 namespace SSW.CleanArchitecture.Domain.UnitTests.Entities;
 
 public class TodoItemTests
 {
     [Fact]
-    public void Should_Be_Creating_Parameterless()
+    public void Create_Should_Succeed_When_Title_Valid()
```
Comment (danielmackay) [inline tests/Domain.UnitTests/Entities/TodoItemTests.cs:9]: Now possible to unit test behavior

File: src/Domain/Entities/TodoItem.cs [hunk 3]
```diff
@@ -1,16 +1,54 @@
...
+        var todoItem = new TodoItem
+        {
+            Id = new TodoItemId(Guid.NewGuid()),
+            Title = title,
+            Priority = PriorityLevel.None,
+            Done = false
+
+        };
+
+        todoItem.AddDomainEvent(new TodoItemCreatedEvent(todoItem));
```
Comment (danielmackay) [inline src/Domain/Entities/TodoItem.cs:33]: Needed to encapsulate creation of `TodoItem` so that we can ensure domain events are always fired.

File: src/Domain/Entities/TodoItem.cs [hunk 4]
```diff
@@ -1,16 +1,54 @@
...
+        todoItem.Reminder = reminder;
+
+        return todoItem;
+    }
+
+    public void Complete()
+    {
+        Done = true;
+
+        AddDomainEvent(new TodoItemCompletedEvent(this));
```
Comment (danielmackay) [inline src/Domain/Entities/TodoItem.cs:52]: Added extra example of encapsulating creation of domain events

File: src/Infrastructure/Persistence/Configuration/TodoItemConfiguration.cs (hunk 1 above)
Comment (christoment) [reply src/Infrastructure/Persistence/Configuration/TodoItemConfiguration.cs:19]: I think it's very rarely you'd want to manually set GUIDs for an entry.

Setting GUID manually is usually needed for application layer logic (I'm thinking of use case where the ID come from external e.g. Auth0). In this case, we just need to create a new field to track this ID, so our domain object entry has the `Id` controlled by our application, and another ID e.g. `auth0Id` coming from external.

That's my opinion - keen to hear other's thoughts 😄

File: src/WebApi/wwwroot/api/specification.json
Comment (christoment) [inline src/WebApi/wwwroot/api/specification.json:1]: Not part of this PR scope - I think we should stop checking in `specification.json` - this should be build-time generated document 🤔

The consequences of this is that if we have other clients generated from this file (e.g. Frontend) then it will have to wait until this build is finished first, but I think that's fine since that's the build pipeline concern 😄

Comment (danielmackay) [reply src/WebApi/wwwroot/api/specification.json:1]: @christoment - can you please create a PBI in the backlog for this?

Comment (christoment) [reply src/WebApi/wwwroot/api/specification.json:1]: Done - https://github.com/SSWConsulting/SSW.CleanArchitecture/issues/122 created 😄

File: src/Infrastructure/Persistence/Configuration/TodoItemConfiguration.cs (hunk 1 above)
Comment (danielmackay) [reply src/Infrastructure/Persistence/Configuration/TodoItemConfiguration.cs]: @christoment fair point.  I've reverted this change, so that we err on the side of simplicity for the template.  Individual projects can always decide to go whichever way they need.

**Comment by danielmackay** [comment]
As discussed with @william-liebenberg - I will add an ADR to cover the reasoning behind using static factory methods.

**Review by william-liebenberg — APPROVED** [review]

**Review by matt-goldman — APPROVED** [review]
LGTM

**Review by christoment — APPROVED** [review]
LGTM 😄
